compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2589,30 +2589,66 @@ impl<'a> Parser<'a> {
         }
     }
 
-    /// Parses `for <src_pat> in <src_expr> <src_loop_block>` (`for` token already eaten).
-    fn parse_expr_for(&mut self, opt_label: Option<Label>, lo: Span) -> PResult<'a, P<Expr>> {
-        // Record whether we are about to parse `for (`.
-        // This is used below for recovery in case of `for ( $stuff ) $block`
-        // in which case we will suggest `for $stuff $block`.
-        let begin_paren = match self.token.kind {
-            token::OpenDelim(Delimiter::Parenthesis) => Some(self.token.span),
-            _ => None,
+    fn parse_for_head(&mut self) -> PResult<'a, (P<Pat>, P<Expr>)> {
+        let pat = if self.token.kind == token::OpenDelim(Delimiter::Parenthesis) {
```

Comment:
instead of branching on this here, can we keep the previous code that saved it into a local and then calls `parse_path_allow_top_alt`? you can then match on the result like in the if case and just return the error if there's no `(`.
that makes it more obvious that the `(` only matters in the error case, which is important

Reply:
I think I got what you meant, check the last commit to verify I did.

Reply:
you did, thanks!